core/services/functions/connector_handler.go

```diff
@@ -160,6 +200,105 @@ func (h *functionsConnectorHandler) handleSecretsSet(ctx context.Context, gatewa
+		})
+		return
+	}
+
+	internalId := InternalId(fromAddr.Bytes(), request.RequestId)
+	request.RequestId = internalId[:]
+	h.lggr.Infow("handling offchain heartbeat", "messageId", requestBody.MessageId, "id", internalId, "sender", requestBody.Sender)
+	h.mu.Lock()
+	response, ok := h.heartbeatRequests[internalId]
+	if !ok { // new request
+		response = &HeartbeatResponse{
+			Status:     RequestStatePending,
+			ReceivedTs: uint64(time.Now().Unix()),
+		}
+		h.cacheNewRequestLocked(internalId, response)
+		h.shutdownWaitGroup.Add(1)
+		go h.handleOffchainRequest(request)
+	}
+	responseToSend := *response
+	h.mu.Unlock()
+	requestBody.Receiver = requestBody.Sender
+	h.sendResponseAndLog(ctx, gatewayId, requestBody, responseToSend)
+}
+
+func (h *functionsConnectorHandler) handleOffchainRequest(request *OffchainRequest) {
+	defer h.shutdownWaitGroup.Done()
+	ctx, cancel := context.WithTimeout(h.serviceContext, time.Duration(HeartbeatRequestTimeoutSec)*time.Second)
+	defer cancel()
+	err := h.listener.HandleOffchainRequest(ctx, request)
+	if err != nil {
+		h.lggr.Infow("internal error while processing", "id", request.RequestId)
+		h.mu.Lock()
+		defer h.mu.Unlock()
+		state, ok := h.heartbeatRequests[RequestID(request.RequestId)]
+		if !ok {
+			h.lggr.Errorw("request unexpectedly disappeared from local cache", "id", request.RequestId)
+			return
+		}
+		state.CompletedTs = uint64(time.Now().Unix())
+		state.Status = RequestStateInternalError
+		state.InternalError = err.Error()
+	} else {
+		// no error - results will be sent to OCR aggregation and returned via reportLoop()
+		h.lggr.Infow("request processed successfully, waiting for aggregation ...", "id", request.RequestId)
+	}
+}
+
+// Listen to OCR reports passed from the plugin and process them against a local cache of requests.
+func (h *functionsConnectorHandler) reportLoop() {
+	defer h.shutdownWaitGroup.Done()
+	for {
+		select {
+		case report := <-h.offchainTransmitter.ReportChannel():
+			h.lggr.Infow("received report", "requestId", report.RequestId, "resultLen", len(report.Result), "errorLen", len(report.Error))
+			if len(report.RequestId) != RequestIDLength {
+				h.lggr.Errorw("report has invalid requestId", "requestId", report.RequestId)
+				continue
+			}
+			h.mu.Lock()
+			cachedResponse, ok := h.heartbeatRequests[RequestID(report.RequestId)]
+			if !ok {
+				h.lggr.Infow("received report for unknown request, caching it", "id", report.RequestId)
+				cachedResponse = &HeartbeatResponse{}
+				h.cacheNewRequestLocked(RequestID(report.RequestId), cachedResponse)
+			}
+			cachedResponse.CompletedTs = uint64(time.Now().Unix())
+			cachedResponse.Status = RequestStateComplete
+			cachedResponse.Response = report
+			h.mu.Unlock()
+		case <-h.chStop:
+			h.lggr.Info("exiting reportLoop")
+			return
+		}
+	}
+}
+
+func (h *functionsConnectorHandler) cacheNewRequestLocked(requestId RequestID, response *HeartbeatResponse) {
+	// remove oldest requests
```

Comment:
As a general note, maybe this "caching" can be simply replaced with a timer-based cleanup? e.g. for a heartbeat you can run `go func() { <- time.After(...); delete(...) }`, maybe with a global atomic counter to prevent flooding.. which you need anyway.

Reply:
UPD: maybe not that simple, worth also watching for `h.serviceContext` being cancelled to release all waiting goroutines. And maybe my approach is more dangerous.

Reply:
Yeah, I'd rather keep it this way for simplicity.